search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,433 @@
+
+G_DEFINE_TYPE_WITH_CODE (EksArticleMetadataProvider,
+                         eks_article_metadata_provider,
+                         G_TYPE_OBJECT,
+                         G_IMPLEMENT_INTERFACE (EKS_TYPE_PROVIDER,
+                                                eks_article_metadata_provider_interface_init))
+
+enum {
+  PROP_0,
+  PROP_APPLICATION_ID,
+  NPROPS
+};
+
+static GParamSpec *eks_article_metadata_provider_props [NPROPS] = { NULL, };
+
+static void
+eks_article_metadata_provider_get_property (GObject    *object,
+                                          unsigned    prop_id,
+                                          GValue     *value,
+                                          GParamSpec *pspec)
+{
+  EksArticleMetadataProvider *self = EKS_ARTICLE_METADATA_PROVIDER (object);
+
+  switch (prop_id)
+    {
+    case PROP_APPLICATION_ID:
+      g_value_set_string (value, self->application_id);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+eks_article_metadata_provider_set_property (GObject      *object,
+                                          unsigned      prop_id,
+                                          const GValue *value,
+                                          GParamSpec   *pspec)
+{
+  EksArticleMetadataProvider *self = EKS_ARTICLE_METADATA_PROVIDER (object);
+
+  switch (prop_id)
+    {
+    case PROP_APPLICATION_ID:
+      g_clear_pointer (&self->application_id, g_free);
+      self->application_id = g_value_dup_string (value);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+eks_article_metadata_provider_finalize (GObject *object)
+{
+  EksArticleMetadataProvider *self = EKS_ARTICLE_METADATA_PROVIDER (object);
+
+  g_clear_pointer (&self->application_id, g_free);
+  g_clear_object (&self->skeleton);
+  g_clear_object (&self->cancellable);
+
+  G_OBJECT_CLASS (eks_article_metadata_provider_parent_class)->finalize (object);
+}
+
+static void
+eks_article_metadata_provider_class_init (EksArticleMetadataProviderClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->get_property = eks_article_metadata_provider_get_property;
+  object_class->set_property = eks_article_metadata_provider_set_property;
+  object_class->finalize = eks_article_metadata_provider_finalize;
+
+  eks_article_metadata_provider_props[PROP_APPLICATION_ID] =
+    g_param_spec_string ("application-id", "Application Id", "Application Id",
+      "", G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class,
+                                     NPROPS,
+                                     eks_article_metadata_provider_props);
+}
+
+typedef struct _ArticleMetadataQueryState {
+  EksArticleMetadataProvider *provider;
+  GDBusMethodInvocation      *invocation;
```

Comment:
I would have expected the state to take ownership of the invocation

Reply:
It probably should.

Reply:
Done